src/Nethermind/Nethermind.Merge.AuRa.Test/ShutterCryptoTests.cs

```diff
@@ -0,0 +1,225 @@
+using System.Linq;
+using Nethermind.Core;
+using Nethermind.Int256;
+using Nethermind.Merge.AuRa.Shutter;
+using NUnit.Framework;
+using Nethermind.Crypto;
+using Nethermind.Core.Extensions;
+using Nethermind.Serialization.Rlp;
+using Nethermind.Core.Test;
+
+namespace Nethermind.Merge.AuRa.Test;
+
+using G1 = Bls.P1;
+using G2 = Bls.P2;
+using GT = Bls.PT;
+using EncryptedMessage = ShutterCrypto.EncryptedMessage;
+
+class ShutterCryptoTests
+{
+    [Test]
+    public void Pairing_holds()
+    {
+        UInt256 sk = 123456789;
+        UInt256 r = 4444444444;
+        G1 identity = G1.generator().mult(3261443);
+        G2 eonKey = G2.generator().mult(sk.ToLittleEndian());
+        G1 key = identity.dup().mult(sk.ToLittleEndian());
+
+        GT p1 = new(key, G2.generator().mult(r.ToLittleEndian()));
+        Bytes32 h1 = ShutterCrypto.Hash2(p1);
+        GT p2 = ShutterCrypto.GTExp(new GT(identity, eonKey), r);
+        Bytes32 h2 = ShutterCrypto.Hash2(p2);
+
+        Assert.That(h1, Is.EqualTo(h2));
+    }
+
+    [Test]
+    [TestCase("f869820243849502f900825208943834a349678ef446bae07e2aeffc01054184af008203e880824fd4a0510c063afbe5b8b8875b680e96a1778c99c765cc0df263f10f8d9707cfa0f114a02590b2ce6dbce6532da17c52a2a7f2eb6155f23404128fca5fb72dc852ce64c6")]
+    [TestCase("08825208943834a349678ef446bae07e2aeffc01054184af008203e880824fd4a02356f869820246849502f900825208943834a349678ef446bae07e2aeffc01054184af008203e880824fd4a02356b138904ed89a72a1fa913aa651c3b4144a5b47aa0cbf6a6cf9956d896bc0a0825208943834a349678ef446bae07e2aeffc01054184af008203e880824fd4a023560825208943834a349678ef446bae07e2aeffc01054184af008203e880824fd4a0235607e1364d24a98ac1cdb3f0af8c5c0cf164528df11dd766aa368d4136651ceb55e")]
+    public void Can_encrypt_then_decrypt(string msgHex)
+    {
+        byte[] msg = Convert.FromHexString(msgHex);
+        UInt256 sk = 123456789;
+        G1 identity = G1.generator().mult(3261443);
+        G2 eonKey = G2.generator().mult(sk.ToLittleEndian());
+        Bytes32 sigma = new([0x12, 0x15, 0xaa, 0xbb, 0x33, 0xfd, 0x66, 0x55, 0x15, 0xaa, 0xbb, 0x33, 0xfd, 0x66, 0x55, 0x15, 0xaa, 0xbb, 0x33, 0xfd, 0x66, 0x55, 0x15, 0xaa, 0xbb, 0x33, 0xfd, 0x66, 0x55, 0x22, 0x88, 0x45]);
+
+        TestContext.WriteLine("eon key for " + sk + ": " + Convert.ToHexString(eonKey.compress()));
+
+        EncryptedMessage encryptedMessage = ShutterCrypto.Encrypt(msg, identity, eonKey, sigma);
+        G1 key = identity.dup().mult(sk.ToLittleEndian());
+
+        Assert.That(ShutterCrypto.RecoverSigma(encryptedMessage, key), Is.EqualTo(sigma));
+        Assert.That(msg.SequenceEqual(ShutterCrypto.Decrypt(encryptedMessage, key)));
+
+        var decoded = ShutterCrypto.DecodeEncryptedMessage(ShutterCrypto.EncodeEncryptedMessage(encryptedMessage));
+        Assert.That(encryptedMessage.C1.is_equal(decoded.C1));
+        Assert.That(encryptedMessage.C2, Is.EqualTo(decoded.C2));
+        Assert.That(encryptedMessage.C3, Is.EqualTo(decoded.C3));
```

Comment:
Can you do 1 assert - can be easily done with Should

Reply:
This fails with a single assert, it can't compare the C1 elements so I have to use `is_equal `